update tidb v4.0.12 release notes (#5090)
pingcap/docs: releases/release-4.0.12.md

---
title: TiDB 4.0.12 Release Notes
---

# TiDB 4.0.12 Release Notes

Release date: March 31, 2021

TiDB version: 4.0.12

## New Features

+ TiFlash

    - Add tools to check exact tiflash replica status for online rolling update [#1512](https://github.com/pingcap/tics/pull/1512)

## Improvements

> 

Reviewer: ```suggestion
+ TiDB
    - Refine explain info for batch cop [#23164](https://github.com/pingcap/tidb/pull/23164)
    - Add warning info for exprs that can not be pushed to storage layer for explain statement. [#23020](https://github.com/pingcap/tidb/pull/23020)
    - Ddl: migrate part of ddl package code from Execute/ExecRestricted to safe API (2) [#22935](https://github.com/pingcap/tidb/pull/22935)
    - Ddl: migrate part of ddl package code from Execute/ExecRestricted to safe API (1). [#22929](https://github.com/pingcap/tidb/pull/22929)
    - Add optimization-time and wait-TS-time into the slow log [#22918](https://github.com/pingcap/tidb/pull/22918)
    - Support query partition_id from infoschema.partitions [#22489](https://github.com/pingcap/tidb/pull/22489)
    - Add 'last_plan_from_binding' to help know whether sql's plan is matched with the hints in the binding [#21430](https://github.com/pingcap/tidb/pull/21430)

```

+ TiFlash

    - Automatically clean archive data to free up disk space [#1638](https://github.com/pingcap/tics/pull/1638)
    - Allow omitting deprecated "users.*" settings in the configuration file of TiFlash. [#1535](https://github.com/pingcap/tics/pull/1535)
    - Reduce the size of TiFlash binary file. [#1414](https://github.com/pingcap/tics/pull/1414)

> 

Reviewer: + TiDB

    - Refine explain info for batch cop [#23164](https://github.com/pingcap/tidb/pull/23164)
    - Add warning info for exprs that can not be pushed to storage layer for explain statement. [#23020](https://github.com/pingcap/tidb/pull/23020)
    - Ddl: migrate part of ddl package code from Execute/ExecRestricted to safe API (2) [#22935](https://github.com/pingcap/tidb/pull/22935)
    - Ddl: migrate part of ddl package code from Execute/ExecRestricted to safe API (1). [#22929](https://github.com/pingcap/tidb/pull/22929)
    - Add optimization-time and wait-TS-time into the slow log [#22918](https://github.com/pingcap/tidb/pull/22918)
    - Support query partition_id from infoschema.partitions [#22489](https://github.com/pingcap/tidb/pull/22489)
    - Add 'last_plan_from_binding' to help know whether sql's plan is matched with the hints in the binding [#21430](https://github.com/pingcap/tidb/pull/21430)
    - Do not report error for prepared stmt execution if tidb_snapshot is set [#22641](https://github.com/pingcap/tidb/pull/22641)
    - Record prepare execute fail as "Failed Query OPM" in monitor [#22672](https://github.com/pingcap/tidb/pull/22672)
    - Add three format specifier for str_to_date expression [#22812](https://github.com/pingcap/tidb/pull/22812)
    - Scattering truncated tables without pre-split option. [#22872](https://github.com/pingcap/tidb/pull/22872)

> 

Reviewer: ```suggestion
+ TiDB

    - Refine explain info for batch cop [#23164](https://github.com/pingcap/tidb/pull/23164)
    - Add warning info for exprs that can not be pushed to storage layer for explain statement. [#23020](https://github.com/pingcap/tidb/pull/23020)
    - Ddl: migrate part of ddl package code from Execute/ExecRestricted to safe API (2) [#22935](https://github.com/pingcap/tidb/pull/22935)
    - Ddl: migrate part of ddl package code from Execute/ExecRestricted to safe API (1). [#22929](https://github.com/pingcap/tidb/pull/22929)
    - Add optimization-time and wait-TS-time into the slow log [#22918](https://github.com/pingcap/tidb/pull/22918)
    - Support query partition_id from infoschema.partitions [#22489](https://github.com/pingcap/tidb/pull/22489)
    - Add 'last_plan_from_binding' to help know whether sql's plan is matched with the hints in the binding [#21430](https://github.com/pingcap/tidb/pull/21430)
    - Do not report error for prepared stmt execution if tidb_snapshot is set [#22641](https://github.com/pingcap/tidb/pull/22641)
    - Record prepare execute fail as "Failed Query OPM" in monitor [#22672](https://github.com/pingcap/tidb/pull/22672)
    - Add three format specifier for str_to_date expression [#22812](https://github.com/pingcap/tidb/pull/22812)
    - Scattering truncated tables without pre-split option. [#22872](https://github.com/pingcap/tidb/pull/22872)
```

    - Add optimization-time and wait-TS-time into the slow log [#22918](https://github.com/pingcap/tidb/pull/22918)
    - Support query partition_id from infoschema.partitions [#22489](https://github.com/pingcap/tidb/pull/22489)
    - Add 'last_plan_from_binding' to help know whether sql's plan is matched with the hints in the binding [#21430](https://github.com/pingcap/tidb/pull/21430)
    - Do not report error for prepared stmt execution if tidb_snapshot is set [#22641](https://github.com/pingcap/tidb/pull/22641)
    - Record prepare execute fail as "Failed Query OPM" in monitor [#22672](https://github.com/pingcap/tidb/pull/22672)
    - Add three format specifier for str_to_date expression [#22812](https://github.com/pingcap/tidb/pull/22812)
    - Scattering truncated tables without pre-split option. [#22872](https://github.com/pingcap/tidb/pull/22872)

> 

Reviewer 2: + TiKV
    - Pd_client: prevent a large number of reconnections in a short time [#9879](https://github.com/tikv/tikv/pull/9879)
    - Optimize seek write for many tombstones [#9729](https://github.com/tikv/tikv/pull/9729)
    - Change the default value of `leader-transfer-max-log-lag` to 128 to increase the success rate of leader transfer [#9605](https://github.com/tikv/tikv/pull/9605)

> + TiFlash

Reviewer 3: ```suggestion
+ PD
    - Save to the region cache when pending-peers or down-peers change [#3471](https://github.com/pingcap/pd/pull/3471)
    - Checker: prevent the regions in split-cache from becoming the target of merge [#3459](https://github.com/pingcap/pd/pull/3459)
+ TiFlash
```

>     - Automatically clean archive data to free up disk space [#1638](https://github.com/pingcap/tics/pull/1638)

Reviewer 4: ```suggestion
```

>     - Optimize configuration file and remove useless items [#1535](https://github.com/pingcap/tics/pull/1535)

Reviewer 4: ```suggestion
    - Optimize configuration file and remove useless items
```

>     - Reduce the size of TiFlash binary file. [#1414](https://github.com/pingcap/tics/pull/1414)

Reviewer 4: ```suggestion
    - Reduce the size of TiFlash binary file
```

>     - Fix get var expr when session var is hex literal [#23372](https://github.com/pingcap/tidb/pull/23372)
>     - Fix the bug that wrong collation is used when try fast path for enum or set [#23292](https://github.com/pingcap/tidb/pull/23292)
>     - Wrong result of nullif expr when used with is null expr [#23279](https://github.com/pingcap/tidb/pull/23279)
>     - Statistics: fix a case that auto-analyze is triggered outside its time range [#23219](https://github.com/pingcap/tidb/pull/23219)
>     - Executor: fix cast function will ignore the error for point get key construction [#23211](https://github.com/pingcap/tidb/pull/23211)
>     - Fixed a bug that prevented SPM from taking effect [#23209](https://github.com/pingcap/tidb/pull/23209)
>     - Fix wrong table filters for index merge plan [#23165](https://github.com/pingcap/tidb/pull/23165)
>     - Fix unexpected NotNullFlag in case when expr ret type [#23135](https://github.com/pingcap/tidb/pull/23135)
>     - Fix a bug that collation is not handle for text type [#23092](https://github.com/pingcap/tidb/pull/23092)
>     - Fix range partition prune bug for IN expr [#23074](https://github.com/pingcap/tidb/pull/23074)
>     - Fix request block forever when replace a tombstone store with the store has same IP addr [#23071](https://github.com/pingcap/tidb/pull/23071)
>     - Scattering truncated tables without pre-split option. [#22872](https://github.com/pingcap/tidb/pull/22872)
>     - Do not ajust int when it is null and compared year [#22844](https://github.com/pingcap/tidb/pull/22844)
>     - Add three format specifier for str_to_date expression [#22812](https://github.com/pingcap/tidb/pull/22812)
>     - Fix load data lost connection error on tables with auto_random column [#22736](https://github.com/pingcap/tidb/pull/22736)
>     - Record prepare execute fail as "Failed Query OPM" in monitor [#22672](https://github.com/pingcap/tidb/pull/22672)
>     - Do not report error for prepared stmt execution if tidb_snapshot is set [#22641](https://github.com/pingcap/tidb/pull/22641)

Reviewer: ```suggestion
    - Fix get var expr when session var is hex literal [#23372](https://github.com/pingcap/tidb/pull/23372)
    - Fix the bug that wrong collation is used when try fast path for enum or set [#23292](https://github.com/pingcap/tidb/pull/23292)
    - Wrong result of nullif expr when used with is null expr [#23279](https://github.com/pingcap/tidb/pull/23279)
    - Statistics: fix a case that auto-analyze is triggered outside its time range [#23219](https://github.com/pingcap/tidb/pull/23219)
    - Executor: fix cast function will ignore the error for point get key construction [#23211](https://github.com/pingcap/tidb/pull/23211)
    - Fixed a bug that prevented SPM from taking effect [#23209](https://github.com/pingcap/tidb/pull/23209)
    - Fix wrong table filters for index merge plan [#23165](https://github.com/pingcap/tidb/pull/23165)
    - Fix unexpected NotNullFlag in case when expr ret type [#23135](https://github.com/pingcap/tidb/pull/23135)
    - Fix a bug that collation is not handle for text type [#23092](https://github.com/pingcap/tidb/pull/23092)
    - Fix range partition prune bug for IN expr [#23074](https://github.com/pingcap/tidb/pull/23074)
    - Fix request block forever when replace a tombstone store with the store has same IP addr [#23071](https://github.com/pingcap/tidb/pull/23071)
    - Do not ajust int when it is null and compared year [#22844](https://github.com/pingcap/tidb/pull/22844)
    - Fix load data lost connection error on tables with auto_random column [#22736](https://github.com/pingcap/tidb/pull/22736)
    - Fix ddl hang over when it meets panic in cancelling path. [#23297](https://github.com/pingcap/tidb/pull/23297)
    - Fix wrong key range of index scan when filter is comparing year column with NULL [#23104](https://github.com/pingcap/tidb/pull/23104)
    - Fix create view success but failed when using it [#23083](https://github.com/pingcap/tidb/pull/23083)
```

> 

Reviewer: ```suggestion
    - Fix ddl hang over when it meets panic in cancelling path. [#23297](https://github.com/pingcap/tidb/pull/23297)
    - Fix wrong key range of index scan when filter is comparing year column with NULL [#23104](https://github.com/pingcap/tidb/pull/23104)
    - Fix create view success but failed when using it [#23083](https://github.com/pingcap/tidb/pull/23083)
```

    - Fix wrong table filters for index merge plan [#23165](https://github.com/pingcap/tidb/pull/23165)
    - Fix unexpected NotNullFlag in case when expr ret type [#23135](https://github.com/pingcap/tidb/pull/23135)
    - Fix a bug that collation is not handle for text type [#23092](https://github.com/pingcap/tidb/pull/23092)
    - Fix range partition prune bug for IN expr [#23074](https://github.com/pingcap/tidb/pull/23074)
    - Fix request block forever when replace a tombstone store with the store has same IP addr [#23071](https://github.com/pingcap/tidb/pull/23071)
    - Do not ajust int when it is null and compared year [#22844](https://github.com/pingcap/tidb/pull/22844)
    - Fix load data lost connection error on tables with auto_random column [#22736](https://github.com/pingcap/tidb/pull/22736)
    - Fix ddl hang over when it meets panic in cancelling path. [#23297](https://github.com/pingcap/tidb/pull/23297)
    - Fix wrong key range of index scan when filter is comparing year column with NULL [#23104](https://github.com/pingcap/tidb/pull/23104)
    - Fix create view success but failed when using it [#23083](https://github.com/pingcap/tidb/pull/23083)

+ PD

>     - Save to the region cache when pending-peers or down-peers change [#3471](https://github.com/pingcap/pd/pull/3471)

Reviewer 3: ```suggestion

```

>     - Checker: prevent the regions in split-cache from becoming the target of merge [#3459](https://github.com/pingcap/pd/pull/3459)

Reviewer 3: ```suggestion
    - Checker: prevent the regions in split-cache from becoming the target of merge [#3459](https://github.com/pingcap/pd/pull/3459)
    - Fix the bug that the isolation level is wrong when the store lacks label [#3474](https://github.com/pingcap/pd/pull/3474)
```

>     - Checker: prevent the regions in split-cache from becoming the target of merge [#3459](https://github.com/pingcap/pd/pull/3459)

Reviewer 3: ```suggestion
```

+ TiFlash

    - Fix an issue of incorrect results when the default value of `binary` type column contains leading or tailing zero bytes
    - Fix the bug that TiFlash fail to sync schema if the name of database contains special characters
    - Fix an issue of incorrect results when handling `IN` expression with decimal values
    - Fix the bug that metrics about opened file count shown in Grafana is high
    - Fix the bug that TiFlash does not support `Timestamp` literal
    - Fix the potential crash while handling `FROM_UNIXTIME` expression
    - Fix an issue of incorrect results when casting string as integer
    - Fix a bug that `like` function may return wrong result

> 

Reviewer 2: + TiKV
    - Fix the issue that the IN expr(coprocessor) didn't handle unsigned/signed int properly [#9850](https://github.com/tikv/tikv/pull/9850)
    - Fix the issue that the ingests operation cannot reentrant. [#9779](https://github.com/tikv/tikv/pull/9779)
    - Fix the issue that the space missed when converting json to string in TiKV coprocessor [#9666](https://github.com/tikv/tikv/pull/9666)

        * BR would log `HTTP_PROXY` and `HTTPS_PROXY` now. [#827](https://github.com/pingcap/br/pull/827)
        * Add cluster_version and br_version info in backupmeta [#803](https://github.com/pingcap/br/pull/803)
        * Check TiDB cluster version before running TiDB-Lightning to avoid unexpected errors. [#787](https://github.com/pingcap/br/pull/787)
        * Improve backup performance when there are many tables. [#745](https://github.com/pingcap/br/pull/745)

    - TiCDC

        * Fix a data loss bug when capture restarts due to network issue, and some table on it is scheduled at the same time. [#1523](https://github.com/pingcap/ticdc/pull/1523)
        * Add double confirm when creating or resuming changefeed with `start-ts` or `checkpoint-ts` 1 day before current ts. [#1499](https://github.com/pingcap/ticdc/pull/1499)
        * No release note, the bug is never released [#1475](https://github.com/pingcap/ticdc/pull/1475)
        * Fix a resolved ts event disorder problems caused by concurrency [#1472](https://github.com/pingcap/ticdc/pull/1472)

>         * No release note, this bug is never released [#1470](https://github.com/pingcap/ticdc/pull/1470)

Reviewer 5: # TiCDC

## Improvements

* Add double confirm when creating or resuming changefeed with start-ts or checkpoint-ts 1 day before current ts. [#1497](https://github.com/pingcap/ticdc/pull/1497)
<!-- * Cli changefeed update will update parameter provided in command only. [#1554](https://github.com/pingcap/ticdc/pull/1554) -->

## Bug Fixes

* Fix a resolved ts event disorder problems caused by concurrency
 [#1464](https://github.com/pingcap/ticdc/pull/1464)
* Fix a data loss bug when capture restarts due to network issue, and some table on it is scheduled at the same time. [#1508](https://github.com/pingcap/ticdc/pull/1508)
<!-- * Fix a bug of untimely release of resources when stopping a processor. [#1547](https://github.com/pingcap/ticdc/pull/1547)
* Fix a bug that txn counter is not correctly updated which may cause database connection leak. [#1524](https://github.com/pingcap/ticdc/pull/1524)
* Fix multiple owner could co-exist when pd has jitter, which may lead to table missing. [#1540](https://github.com/pingcap/ticdc/pull/1540) -->

# TiDB-binlog

None

# BR

## Improvements

* Log `HTTP_PROXY` and `HTTPS_PROXY`. [#827](https://github.com/pingcap/br/pull/827)
* Improve backup performance when there are many tables. [#745](https://github.com/pingcap/br/pull/745)
* Report error if service safe point check fails. [#826](https://github.com/pingcap/br/pull/826)
* Add cluster_version and br_version info in backupmeta [#803](https://github.com/pingcap/br/pull/803)
* Added retry for external storage errors. [#851](https://github.com/pingcap/br/pull/851)
* Reduce memory usage during backup. [#886](https://github.com/pingcap/br/pull/886)

## Bug Fixes

* Fix the bug that WalkDir for s3 storage returns nil if the target path is bucket name. [#773](https://github.com/pingcap/br/pull/773)
* Fixed a bug that caused, even BR started with TLS config, the pprof endpoints won't be served with TLS. [#839](https://github.com/pingcap/br/pull/839)

# Dumpling

None

# TiDB-Lightning

## Improvements

* Check TiDB cluster version before running TiDB-Lightning to avoid unexpected errors. [#787](https://github.com/pingcap/br/pull/787)
* Fail fast when lightning meets error in engine/chunk restore and do not retry context.Cancel error. [#867](https://github.com/pingcap/br/pull/867)
* Added configurations tikv-importer.engine-mem-cache-size and tikv-importer.local-writer-mem-cache-size to tune between memory usage and performance. [#866](https://github.com/pingcap/br/pull/866)
* Run batch split regions in parallel for Lightning local backend. [#868](https://github.com/pingcap/br/pull/868)
* When using Lightning to import from S3, Lightning no longer require s3:ListBucket permission on the entire bucket, only the data source prefix itself. [#919](https://github.com/pingcap/br/pull/919)
* When resuming from checkpoint, Lightning now keeps restoring the original engines instead of starting to process some random new ones. [#924](https://github.com/pingcap/br/pull/924)

## Bug Fixes

* Fix a bug that importer may ignore write rows error if open engine returns file exists error. [#848](https://github.com/pingcap/br/pull/848)
* Fix a bug that lightning generated ts may be to large or small that query may return incorrect result. [#850](https://github.com/pingcap/br/pull/850)
* Fix a bug that lightning unexpected exit may cause checkpoint file truncated to 0 sized. [#889](https://github.com/pingcap/br/pull/889)
* Fix a bug that chunk restore task may ignore context cancel error and causes data loss. [#874](https://github.com/pingcap/br/pull/874)

---
title: TiDB 4.0.12 Release Notes
---

# TiDB 4.0.12 Release Notes

Release date: April 2, 2021

TiDB version: 4.0.12

## New Features

+ TiFlash

>     - Add tools to check the exact status of `tiflash replica` for online rolling update

Reviewer 6: ```suggestion
    - Add tools to check the status of `tiflash replica` for online rolling update
```

## Improvements

+ TiDB

>     - Refine explain info for batch cop [#23164](https://github.com/pingcap/tidb/pull/23164)

Reviewer 6: ```suggestion
    - Refine the output information of the `EXPLAIN` statement in the `batch cop` mode [#23164](https://github.com/pingcap/tidb/pull/23164)
```

>     - Add the warning information for expressions that cannot be pushed to the storage layer in `EXPLAIN` statement [#23020](https://github.com/pingcap/tidb/pull/23020)

Reviewer 6: ```suggestion
    - Add the warning information for expressions that cannot be pushed to the storage layer in the output of the `EXPLAIN` statement [#23020](https://github.com/pingcap/tidb/pull/23020)
```

    - Migrate a part of the DDL package code from `Execute`/`ExecRestricted` to the safe API (2) [#22935](https://github.com/pingcap/tidb/pull/22935)
    - Migrate a part of the DDL package code from `Execute`/`ExecRestricted` to the safe API (1) [#22929](https://github.com/pingcap/tidb/pull/22929)
    - Add `optimization-time` and `wait-TS-time` into the slow log [#22918](https://github.com/pingcap/tidb/pull/22918)
    - Support querying `partition_id` from the `infoschema.partitions` table [#22489](https://github.com/pingcap/tidb/pull/22489)

>     - Add `last_plan_from_binding` to help the user know whether a SQL statement's execution plan is matched with the hints in the binding [#21430](https://github.com/pingcap/tidb/pull/21430)

Reviewer 6: ```suggestion
    - Add `last_plan_from_binding` to help the users know whether a SQL statement's execution plan is matched with the hints in the binding [#21430](https://github.com/pingcap/tidb/pull/21430)
```

    - Scatter truncated tables without the `pre-split` option [#22872](https://github.com/pingcap/tidb/pull/22872)

>     - Add three format specifier for the `str_to_date` expression [#22812](https://github.com/pingcap/tidb/pull/22812)

Reviewer 6: ```suggestion
    - Add three format specifiers for the `str_to_date` expression [#22812](https://github.com/pingcap/tidb/pull/22812)
```

    - Record the `PREPARE` execution failure as `Failed Query OPM` in the metrics monitor [#22672](https://github.com/pingcap/tidb/pull/22672)
    - Do not report errors for the `PREPARE` execution if `tidb_snapshot` is set [#22641](https://github.com/pingcap/tidb/pull/22641)

+ TiKV

    - Prevent a large number of reconnections in a short period of time [#9879](https://github.com/tikv/tikv/pull/9879)

>     - Optimize the write operations in the scenarios of many tombstones [#9729](https://github.com/tikv/tikv/pull/9729)

Reviewer 7: ```suggestion
    - Optimize the write operations and Batch Get in the scenarios of many tombstones [#9729](https://github.com/tikv/tikv/pull/9729)
```

    - Change the default value of `leader-transfer-max-log-lag` to `128` to increase the success rate of leader transfer [#9605](https://github.com/tikv/tikv/pull/9605)

+ PD

>     - Update the Region cache only when `pending-peers` or `down-peers` change, which reduces the pressure of updating heartbeats [#3471](https://github.com/pingcap/pd/pull/3471)

Reviewer 6: ```suggestion
    - Update the Region cache only when `pending-peers` or `down-peers` changes, which reduces the pressure of updating heartbeats [#3471](https://github.com/pingcap/pd/pull/3471)
```

    - Prevent the Regions in `split-cache` from becoming the target of merge [#3459](https://github.com/pingcap/pd/pull/3459)

+ TiFlash

    - Optimize the configuration file and remove useless items
    - Reduce the size of TiFlash binary files
    - Use an adaptive aggressive GC strategy to reduce memory usage

+ Tools

    + TiCDC

>         - Add a double confirmation when creating or resuming the changefeed with the `start-ts` or `checkpoint-ts` 1 day before the current timestamp [#1497](https://github.com/pingcap/ticdc/pull/1497)

Reviewer 6: ```suggestion
        - Add a double confirmation when users create or resume the changefeed with the `start-ts` or `checkpoint-ts` 1 day before the current timestamp [#1497](https://github.com/pingcap/ticdc/pull/1497)
```

    + Backup & Restore (BR)

        - Log the `HTTP_PROXY` and `HTTPS_PROXY` environmental variables [#827](https://github.com/pingcap/br/pull/827)

>         - Improve backup performance when there are many tables [#745](https://github.com/pingcap/br/pull/745)

Reviewer 6: ```suggestion
        - Improve the backup performance when there are many tables [#745](https://github.com/pingcap/br/pull/745)
```

        - Report errors if the service safe point check fails [#826](https://github.com/pingcap/br/pull/826)
        - Add the `cluster_version` and `br_version` information in `backupmeta` [#803](https://github.com/pingcap/br/pull/803)
        - Add retry for external storage errors to increase the success rate of backup [#851](https://github.com/pingcap/br/pull/851)
        - Reduce memory usage during backup [#886](https://github.com/pingcap/br/pull/886)

    + TiDB Lightning

        - Check the TiDB cluster version before running TiDB Lightning to avoid unexpected errors [#787](https://github.com/pingcap/br/pull/787)
        - Fail fast when TiDB Lightning meets the `cancel` error [#867](https://github.com/pingcap/br/pull/867)
        - Add `tikv-importer.engine-mem-cache-size` and `tikv-importer.local-writer-mem-cache-size` configuration items to balance between memory usage and performance [#866](https://github.com/pingcap/br/pull/866)

>         - Run `batch split region` in parallel for TiDB Lightning's Local-backend to increase import speed [#868](https://github.com/pingcap/br/pull/868)

Reviewer 6: ```suggestion
        - Run `batch split region` in parallel for TiDB Lightning's Local-backend to increase the import speed [#868](https://github.com/pingcap/br/pull/868)
```

    - Do not adjust the `INT` type when it is `NULL` and compared with `YEAR` [#22844](https://github.com/pingcap/tidb/pull/22844)
    - Fix the issue of lost connection when loading data on tables with the `auto_random` column [#22736](https://github.com/pingcap/tidb/pull/22736)
    - Fix the issue of DDL hangover when the DDL operation meets panic in the cancelling path [#23297](https://github.com/pingcap/tidb/pull/23297)
    - Fix the wrong key range of index scan when comparing the `YEAR` column with `NULL` [#23104](https://github.com/pingcap/tidb/pull/23104)
    - Fix the issue that a successfully created view is failed to use [#23083](https://github.com/pingcap/tidb/pull/23083)

+ TiKV

    - Fix the issue that the `IN` expression does not properly handle unsigned/signed integers [#9850](https://github.com/tikv/tikv/pull/9850)
    - Fix the issue that the ingest operation is not re-entrant [#9779](https://github.com/tikv/tikv/pull/9779)
    - Fix the issue that the space is missed when converting JSON to string in TiKV coprocessor [#9666](https://github.com/tikv/tikv/pull/9666)

+ PD

>     - Fix the bug that the isolation level is wrong when the store lacks the label [#3474](https://github.com/pingcap/pd/pull/3474)

Reviewer 6: ```suggestion
    - Fix a bug that the isolation level is wrong when the store lacks the label [#3474](https://github.com/pingcap/pd/pull/3474)
```

+ TiFlash

>     - Fix an issue of incorrect execution results when the default value of the `binary` type column contains leading or tailing zero bytes

Reviewer 6: ```suggestion
    - Fix the issue of incorrect execution results when the default value of the `binary` type column contains leading or tailing zero bytes
```

>     - Fix the bug that TiFlash fails to synchronize schema if the name of the database contains special characters

Reviewer 6: ```suggestion
    - Fix a bug that TiFlash fails to synchronize schema if the name of the database contains special characters
```

>     - Fix the bug that TiFlash fails to synchronize schema if the name of the database contains special characters

Reviewer 6: ```suggestion
    - Fix a bug that TiFlash fails to synchronize schema if the name of the database contains special characters
```

>     - Fix an issue of incorrect results when handling the `IN` expression with decimal values

Reviewer 6: ```suggestion
    - Fix the issue of incorrect results when handling the `IN` expression with decimal values
```

    - Fix a bug that the metric for the opened file count shown in Grafana is high
    - Fix a bug that TiFlash does not support the `Timestamp` literal

>     - Fix the potential crash while handling `FROM_UNIXTIME` expression

Reviewer 6: Suggest not using the word "crash" in the user doc.
```suggestion
    - Fix the potential not responding issue while handling `FROM_UNIXTIME` expression
```

>     - Fix an issue of incorrect results when casting string as integer

Reviewer 6: ```suggestion
    - Fix the issue of incorrect results when casting string as integer
```

    - Fix a bug that the `like` function might return wrong results

+ Tools

    + TiCDC

        - Fix a disorder issue of the `resolved ts` event [#1464](https://github.com/pingcap/ticdc/pull/1464)

>         - Fix a data loss issue caused by wrong table scheduling due to the network problem [#1508](https://github.com/pingcap/ticdc/pull/1508)

Reviewer 5: Please hold, there are some pending fixes that need to be added.

Reviewer 5: ```suggestion
        - Fix a data loss issue caused by wrong table scheduling due to the network problem [#1508](https://github.com/pingcap/ticdc/pull/1508)
        - Fix A bug of untimely release of resources when stopping a processor [#1547](https://github.com/pingcap/ticdc/pull/1547)
        - Fix a bug that txn counter is not correctly updated which may cause database connection leak [#1524](https://github.com/pingcap/ticdc/pull/1524)
        - Fix multiple owner could co-exist when pd has jitter, which may lead to table missing [#1540](https://github.com/pingcap/ticdc/pull/1540)
```

    + Backup & Restore (BR)

>         - Fix a bug that `WalkDir` for the s3 storage returns `nil` if the target path is bucket name [#773](https://github.com/pingcap/br/pull/773)

Reviewer 6: The PR number for this bug fix item is wrong. PTAL.
```suggestion
        - Fix a bug that `WalkDir` for the s3 storage returns `nil` if the target path is bucket name [#733](https://github.com/pingcap/br/pull/733)
```

+ TiFlash

    - Allow omitting deprecated "users.*" settings in the configuration file of TiFlash [#1535](https://github.com/pingcap/tics/pull/1535)
    - Reduce the size of TiFlash binary file [#1414](https://github.com/pingcap/tics/pull/1414)

## Bug Fixes

+ PD

    - Save to the region cache when pending-peers or down-peers change [#3471](https://github.com/pingcap/pd/pull/3471)
    - Checker: prevent the regions in split-cache from becoming the target of merge [#3459](https://github.com/pingcap/pd/pull/3459)

+ TiFlash

    - Fix the issue that binary type column's default value containing leading zero bytes is trimmed and does not contain padded tailing zero bytes [#1571](https://github.com/pingcap/tics/pull/1571)
    - Fix the bug that TiFlash fail to sync schema if the database name contains special character [#1559](https://github.com/pingcap/tics/pull/1559)
    - Use an adaptive aggressive GC strategy to avoid OOM [#1553](https://github.com/pingcap/tics/pull/1553)
    - Fix wrong result of IN function [#1532](https://github.com/pingcap/tics/pull/1532)
    - Fix the bug that opened file count shown in Grafana is high [#1503](https://github.com/pingcap/tics/pull/1503)

>     - Support timestamp literal in dag requeset [#1501](https://github.com/pingcap/tics/pull/1501)

Reviewer 4: ```suggestion
    - Fix the issue that TiFlash does not support `Timestamp` literal
```

>     - Fix potential segmentation fault in TiFlash [#1491](https://github.com/pingcap/tics/pull/1491)

Reviewer 4: ```suggestion
    - Fix the potential crash while handling `FROM_UNIXTIME` expression
```

>     - Fix problem CastStringAsInt may produce wrong result. [#1481](https://github.com/pingcap/tics/pull/1481)

Reviewer 4: ```suggestion
    - Fix an issue of incorrect results when casting string as integer
```

>     - Fix a bug that the `like` function may return wrong result [#1461](https://github.com/pingcap/tics/pull/1461)

Reviewer 4: ```suggestion
    - Fix a bug that `like` function may return wrong result
```
